### PR TITLE
Add status report support for IM read

### LINKED ISSUE #9191: Add status report support in IM read
#### Problem
Per Interacton model spec, we need to implement status report in response to report data during read.
"Data reports (i.e. non-empty ReportData messages) SHALL always require a response. In those cases, the SuppressResponse flag isn’t applicable and SHALL not be sen"
https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/data_model/Encoding-Specification.adoc#reportdatamessage

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
* Fixes #9191
#### Change overview
-- When Read Client receives report data from Read handler, it needs to
send status report to handler per IM spec("Data reports (i.e. non-empty ReportData messages) SHALL always require a response."
https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/data_model/Encoding-Specification.adoc#reportdatamessage)
-- Add status report generation and process.
-- Add corresponding unit test and integration test

#### Testing
How was this tested? (at least one bullet point required)
Update the unit and integration test to check if status report is successfully sent.

### CODE REVIEW

File: src/app/ReadHandler.cpp [hunk 1]
```diff
@@ -75,24 +78,104 @@ CHIP_ERROR ReadHandler::OnReadRequest(Messaging::ExchangeContext * apExchangeCon
...
+}
+
+CHIP_ERROR ReadHandler::OnMessageReceived(Messaging::ExchangeContext * apExchangeContext, const PacketHeader & aPacketHeader,
+                                          const PayloadHeader & aPayloadHeader, System::PacketBufferHandle && aPayload)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    if (aPayloadHeader.HasMessageType(Protocols::SecureChannel::MsgType::StatusReport))
+    {
+        ChipLogProgress(DataManagement, "Receive Status Report");
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/ReadHandler.cpp]: Done, just make it more useful with printing protocol code and protocol code,  thanks

**Comment by woody-apple** [comment]
@vivien-apple @bzbarsky-apple ?

**Review by woody-apple — DISMISSED** [review]
How does this tie into existing ember reporting and how we're adapting it to be spec compliant for subscriptions?

**Comment by andy31415** [comment]
> How does this tie into existing ember reporting and how we're adapting it to be spec compliant for subscriptions?

@yunhanw-google - could you provide the spec link that provides details with what we are looking to implement.

@woody-apple the way I read is that this is for IM reads so it should not be related to ember reporting but rather IM reads.

**Comment by mrjerryjohns** [comment]
@woody-apple Existing Ember reporting is only about subscriptions. This PR adds support for status reports for report generation on Reads only, which is currently managed by the Read reporting engine.

This implements sections 10.2.4.1 in the Encoding specification, which provides the specific message flow that is being implemented here (i.e a StatusReport sent in response to a Read request).

File: src/app/ReadClient.cpp [hunk 2]
```diff
@@ -226,19 +265,13 @@ CHIP_ERROR ReadClient::OnMessageReceived(Messaging::ExchangeContext * apExchange
...
-                 err = CHIP_ERROR_INVALID_MESSAGE_TYPE);
-    err = ProcessReportData(std::move(aPayload));
-
-exit:
-    ChipLogFunctError(err);
-
-    MoveToState(ClientState::Initialized);
-
-    if (mpDelegate != nullptr)
+    if (aPayloadHeader.HasMessageType(Protocols::InteractionModel::MsgType::ReportData))
```
Comment (andy31415) [inline src/app/ReadClient.cpp:271]: nit: why not use the verify macro  here? That would decrease the indentation and make it obvious that on non-report-data we return INVALID_MESSAGE_TYPE.

**Review by andy31415 — APPROVED** [review]

File: src/app/ReadClient.cpp (hunk 2 above)
Comment (yunhanw-google) [reply src/app/ReadClient.cpp]: Done, thanks

**Comment by woody-apple** [comment]
Removing "SDK Discussion Required" as this appears to be an incremental addition to read subscriptions.

**Review by jmartinez-silabs — APPROVED** [review]

**Review by jelderton — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

**Pull request merged**